### PR TITLE
rework stack management

### PR DESCRIPTION
**Pull request by errordeveloper**
### Description

- cater for dry-run
- cleaner main functions for create and delete commands
- more general abstraction for tasks
- expose all common operations through functions
- remove special purpose functions

### Checklist

- [x] Code compiles correctly (i.e `make build`)
- [x] Added tests that cover your change (if possible)
- [x] All unit tests passing (i.e. `make test`)
- [x] All integration tests passing (i.e. `make integration-test`)
- [x] Added/modified documentation as required (such as the `README.md`, and `examples` directory)

### CODE REVIEW

**Comment by errordeveloper** [comment]
@rndstr would you mind to take a look at this please?

**Comment by errordeveloper** [comment]
Still need to rewrite the unit tests.

**Comment by errordeveloper** [comment]
I think there is a bug, like a possible race condition... I'll need to investigate, perhaps unit tests will show.

File: pkg/ctl/delete/cluster.go [hunk 1]
```diff
@@ -76,75 +100,46 @@ func doDeleteCluster(p *api.ProviderConfig, cfg *api.ClusterConfig, nameArg stri
...
-					logger.Critical("%s\n", err.Error())
-				}
-				return fmt.Errorf("failed to delete nodegroup(s)")
-			}
-			return nil
-		}
-		if err := tryDeleteAllNodeGroups(false); err != nil {
-			logger.Info("will retry deleting nodegroup")
+		tasks, err := stackManager.DeleteTasksForClusterWithNodeGroups(wait, func(_ chan error, _ string) error {
 			logger.Info("trying to cleanup dangling network interfaces")
```
Comment (rndstr) [inline pkg/ctl/delete/cluster.go]: this cleanup function is being called once for each nodegroup now. it looks like it was intended to be just run once previously?

File: pkg/cfn/manager/tasks.go [hunk 2]
```diff
@@ -1,158 +1,208 @@
...
+	stack *Stack
+	call  func(*Stack, chan error) error
+}
+
+func (t *taskWithStackSpec) Describe() string { return t.info }
+func (t *taskWithStackSpec) Do(errs chan error) error {
+	return t.call(t.stack, errs)
+}
+
+type asyncTaskWithStackSpec struct {
```
Comment (rndstr) [inline pkg/cfn/manager/tasks.go]: can you elaborate what the purpose for this struct is and what is intended to happen when the error channel is closed in `Do()`?

File: pkg/cfn/manager/deprecated.go [hunk 3]
```diff
@@ -1,57 +1,98 @@
...
-	return err
+	cpStackFound := false
+	for _, s := range stacks {
+		if strings.HasSuffix(*s.StackName, "-ControlPlane") {
+			cpStackFound = true
+		}
+	}
+	tasks := &TaskSet{}
+
+	for _, suffix := range deprecastedStackSuffices() {
```
Comment (martina-if) [inline pkg/cfn/manager/deprecated.go]: Typo: ~deprecastedStackSuffices~ -> deprecatedStackSuffices

File: pkg/cfn/manager/tasks.go [hunk 4]
```diff
@@ -1,158 +1,208 @@
...
-		if onlySubset != nil && !onlySubset.Has(ng.Name) {
-			continue
-		}
-		t := Task{
-			Call: c.CreateNodeGroup,
-			Data: ng,
-		}
-		createAllNodeGroups = append(createAllNodeGroups, t)
+	if t.Parallel {
+		doParallelTasks(appendErr, t.tasks)
```
Comment (martina-if) [inline pkg/cfn/manager/tasks.go]: I think there is still a race condition here: there will be several goroutines in parallel appending to the `errs` slice, which I think is not thread safe. I think `doParallelTasks()` should collect the errors through channels, then combine them and return them. (Or use a mutex, which I think may look uglier)

Comment (errordeveloper) [reply pkg/cfn/manager/tasks.go]: I think you are right, I'm seeing misbehaviour in some of my tests. I was hoping unit tests will bring it to justice! :)
It sounds like a channel here would help, but let me just write up some unit tests first.

File: pkg/cfn/manager/tasks.go [hunk 5]
```diff
@@ -1,158 +1,208 @@
...
+	Do(chan error) error
+	Describe() string
+}
 
-// Task has a function with an opaque payload
-type Task struct {
-	Call taskFunc
-	Data interface{}
+// TaskSet wraps a set of tasks
+type TaskSet struct {
```
[the remark this answers is not in the document]
Comment (errordeveloper) [reply pkg/cfn/manager/tasks.go]: Yeah, good point, thanks! I just called it somethings, cause `Tasks` would be confusing, given we have `Task` already.

File: pkg/ctl/delete/cluster.go (hunk 1 above)
Comment (errordeveloper) [reply pkg/ctl/delete/cluster.go]: Yeah, that's true, but I think it'd be okay for now, we need to do more here anyway (see #536). It shouldn't hurt to call this one multiple times.

File: pkg/cfn/manager/tasks.go (hunk 2 above)
Comment (errordeveloper) [reply pkg/cfn/manager/tasks.go]: So the error channel is closed here right after the handler returns, to indicate the task has completed. Maybe we could give this one a better name... It's a fire-and-forget type of task.

File: pkg/cfn/manager/deprecated.go (hunk 3 above)
Comment (errordeveloper) [reply pkg/cfn/manager/deprecated.go]: 🤣

File: pkg/cfn/manager/delete_tasks.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// DeleteTasksForClusterWithNodeGroups defines tasks required to delete all the nodegroup
+// stacks and the cluster
+func (c *StackCollection) DeleteTasksForClusterWithNodeGroups(wait bool, cleanup func(chan error, string) error) (*TaskSet, error) {
```
[the remark this answers is not in the document]
Comment (errordeveloper) [reply pkg/cfn/manager/delete_tasks.go]: How about `NewTasksToDeleteNodeGroups`, `NewTasksToDeleteCluster` etc?

**Comment by errordeveloper** [comment]
@martina-if thanks for spotting the race condition! As I mentioned, I was able to reproduce it with unit tests. I've fixed it a we have quite a few tests now. This PR is in a pretty good shape to merge next week :)

File: pkg/cfn/manager/tasks.go [hunk 7]
```diff
@@ -1,158 +1,233 @@
...
-	for _, s := range nodeGroupStacks {
-		if err := c.DeleteNodeGroup(getNodeGroupName(s)); err != nil {
-			errs = append(errs, err)
-		}
+	allErrs := []error{}
+	for err := range errs {
+		allErrs = append(allErrs, err)
 	}
+	return allErrs
+	// for {
```
Comment (errordeveloper) [inline pkg/cfn/manager/tasks.go]: @rndstr would you happen to know if `for x := range <channel>` is exactly the same as above? It seems like it is, but I'm not 100% sure.

File: pkg/cfn/manager/tasks_test.go [hunk 8]
```diff
@@ -1,104 +1,328 @@
...
 
-					return nil
+					Expect(tasks.Describe()).To(Equal("2 sequential tasks: { 2 sequential sub-tasks: { t1.1, 2 parallel sub-tasks: { t3.1, t3.2 } }, t2.1 }"))
+
+					errs := tasks.DoAllSync()
+					Expect(errs).To(HaveLen(2))
+					Expect(errs[0].Error()).To(Equal("t3.2 always fails"))
+					Expect(errs[1].Error()).To(Equal("t2.1 always fails"))
+
+					Expect(called11).To(BeTrue())
```
Comment (errordeveloper) [inline pkg/cfn/manager/tasks_test.go]: This checks if tasks have been called, but it'd be nice to test the effective order of completion... I was wondering would be a good way to test that? Perhaps a channel or thread-safe slice with a mutex?

File: pkg/cfn/manager/tasks_test.go [hunk 9]
```diff
@@ -1,104 +1,328 @@
...
+
+					tasks.DryRun = true
+
+					Expect(tasks.DoAllSync()).To(HaveLen(0))
+
+					tasks.DryRun = false
+					errs := tasks.DoAllSync()
+					Expect(errs).To(HaveLen(2))
+					Expect(errs[0].Error()).To(Equal("t1.1 always fails"))
+					Expect(errs[1].Error()).To(Equal("t1.3 always fails"))
```
Comment (errordeveloper) [inline pkg/cfn/manager/tasks_test.go]: Should add another case where a task return an error right way, which should also come on the channel...

File: pkg/cfn/manager/tasks.go (hunk 7 above)
Comment (rndstr) [reply pkg/cfn/manager/tasks.go]: yep looks the same to me.
(with the caveat that if you were to have code after that block you'd need a `break <label>` to get out of the for _and_ select  I think)

Comment (errordeveloper) [reply pkg/cfn/manager/tasks.go]: Yeah... I think select just let's you handle multiple channels nicely, it makes sense to have `for/range` shorthand syntax for a single channel.

**Comment by errordeveloper** [comment]
Looks like we cannot use `go test -race` in alpine container due to CGO/musl:
```
# runtime/race
/usr/lib/gcc/x86_64-alpine-linux-musl/8.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: race_linux_amd64.syso: in function `__sanitizer::GetArgv()':
gotsan.cc:(.text+0x4183): undefined reference to `__libc_stack_end'
/usr/lib/gcc/x86_64-alpine-linux-musl/8.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: race_linux_amd64.syso: in function `__sanitizer::ReExec()':
gotsan.cc:(.text+0x9797): undefined reference to `__libc_stack_end'
/usr/lib/gcc/x86_64-alpine-linux-musl/8.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: race_linux_amd64.syso: in function `__sanitizer::InternalAlloc(unsigned long, __sanitizer::SizeClassAllocatorLocalCache<__sanitizer::SizeClassAllocator32<__sanitizer::AP32> >*, unsigned long)':
gotsan.cc:(.text+0xaac1): undefined reference to `__libc_malloc'
/usr/lib/gcc/x86_64-alpine-linux-musl/8.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: race_linux_amd64.syso: in function `__sanitizer::InternalRealloc(void*, unsigned long, __sanitizer::SizeClassAllocatorLocalCache<__sanitizer::SizeClassAllocator32<__sanitizer::AP32> >*)':
gotsan.cc:(.text+0xca20): undefined reference to `__libc_realloc'
/usr/lib/gcc/x86_64-alpine-linux-musl/8.2.0/../../../../x86_64-alpine-linux-musl/bin/ld: race_linux_amd64.syso: in function `__sanitizer::InternalFree(void*, __sanitizer::SizeClassAllocatorLocalCache<__sanitizer::SizeClassAllocator32<__sanitizer::AP32> >*)':
gotsan.cc:(.text+0x66e8): undefined reference to `__libc_free'
collect2: error: ld returned 1 exit status
```

File: pkg/cfn/manager/tasks_test.go (hunk 8 above)
Comment (errordeveloper) [reply pkg/cfn/manager/tasks_test.go]: Tried a channel, and it didn't work too well. Switched to a slice with a mutex, seems to work fine.

File: pkg/cfn/manager/delete_tasks.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (errordeveloper) [reply pkg/cfn/manager/delete_tasks.go]: I've renamed these to `NewTasksToDeleteX` and `NewTasksToCreateX`, let me know how this looks!

File: pkg/cfn/manager/tasks.go [hunk 10]
```diff
@@ -1,158 +1,215 @@
...
-// Task has a function with an opaque payload
-type Task struct {
-	Call taskFunc
-	Data interface{}
+// TaskTree wraps a set of tasks
+type TaskTree struct {
+	tasks    []Task
+	Parallel bool
+	DryRun   bool
+	Sub      bool
```
Comment (martina-if) [inline pkg/cfn/manager/tasks.go]: What is Sub?

File: pkg/cfn/manager/tasks.go [hunk 11]
```diff
@@ -1,158 +1,215 @@
...
+func (t *asyncTaskWithStackSpec) Do(errs chan error) error {
+	_, err := t.call(t.stack)
+	close(errs)
+	return err
+}
+
+func doSingleTask(allErrs chan error, task Task) {
+	desc := task.Describe()
+	logger.Debug("started task: %s", desc)
+	errs := make(chan error)
```
Comment (martina-if) [inline pkg/cfn/manager/tasks.go]: Why not share the same allErrs channel?

Comment (martina-if) [reply pkg/cfn/manager/tasks.go]: defer errs.close()?

File: pkg/cfn/manager/tasks.go (hunk 10 above)
Comment (errordeveloper) [reply pkg/cfn/manager/tasks.go]: A sub-task, i.e. child not a parent. It's just so we can print user-friendly messages. Would `Child` be better? Besides, please do review the message formatting, it should evident from the tests and the readme changes.

File: pkg/cfn/manager/tasks.go (hunk 11 above)
Comment (errordeveloper) [reply pkg/cfn/manager/tasks.go]: > Why not share the same allErrs channel?

The task itself is expected to close this channel, and what we see on this channel gets copied into shared `allErrs` channel. I don't think we should change that now.

File: pkg/cfn/manager/api.go [hunk 12]
```diff
@@ -243,38 +245,55 @@ func (c *StackCollection) DeleteStack(name string, force bool) (*Stack, error) {
...
-		return nil, fmt.Errorf("stack %q previously couldn't be deleted", name)
-	}
 	i.StackId = s.StackId
+	return c.DeleteStackBySpec(i)
+}
+
+// WaitDeleteStackByName sends a request to delete the stack, and waits until status is DELETE_COMPLETE;
+// any errors will be written to errs channel, assume completion when nil is written, do not expect
+// more then one error value on the channel, it's closed immediately after it is written to
+func (c *StackCollection) WaitDeleteStackByName(name string, errs chan error) error {
```
[the remark this answers is not in the document]
Comment (errordeveloper) [reply pkg/cfn/manager/api.go]: 👍

File: pkg/cfn/manager/tasks.go (hunk 10 above)
Comment (martina-if) [reply pkg/cfn/manager/tasks.go]: maybe something like `isChild`?

Comment (martina-if) [reply pkg/cfn/manager/tasks.go]: or isSubtask?

Comment (errordeveloper) [reply pkg/cfn/manager/tasks.go]: Let's go with `isSubtask`, cause that doesn't imply changing words that are used in output.

**Pull request merged**